woocommerce/woocommerce-ios
# Store creation M2: integrate with mock IAP to test the entire flow

Part of #8108


### Description


#### Using an IAP mock

There is still an IAP backend issue (p1668419146340439-slack-C025A8VV728) that results in a `Server could not read response` failure (p1668592527610099/1668419146.340439-slack-C025A8VV728) when I tested purchasing the eCommerce plan. In order to unblock us from testing and developing the store creation flow, I decided to create a mock implementation of `InAppPurchasesForWPComPlansProtocol` in debug builds using the DEBUG macro and a feature flag `storeCreationM2WithInAppPurchasesEnabled` to easily turn it on and off. The DEBUG macro is used so that the mock struct isn't included in non-debug builds, since it's generally a good practice not to include mock/debug classes in production builds. If you feel like the use of the macro makes it harder to read, I can also remove the macro condition since it's only temporary.

#### Changes to navigation bar title & back button styles

Because the `store name form` (`StoreNameForm`) and `plan purchase view` (`StoreCreationPlanView`) screens do not use any navigation bar title, before this PR a gap is shown below the navigation bar for these screens. I struggled with various navigation bar title styles in all different screens in the flow, sometimes disabling the large title style in one screen results in the next screen not being able to show a large title. I also got stuck on hiding the back button in the navigation bar for `StoreCreationPlanView`. After debugging for a while, I realized that the issues were from having SwiftUI view in a `UIHostingController` and setting navigation bar title and back button in the hosting controller 🤦🏻‍♀️ As a result, I changed the navigation bar title and back button configurations from the hosting controller to the SwiftUI view.

#### Changes to `StoreCreationCoordinator`

Using the IAP interface `InAppPurchasesForWPComPlansProtocol`, we can now integrate the whole flow with each step in place. To wait for the newly created site to become a Jetpack site, I copied most of the logic from M1 implementation but the site ID subscription feels a bit odd to me since it never changes. Please feel free to suggest a better solution not using Combine.

### Testing instructions


- Log in if needed
- Go to the Menu tab, and tap `Switch store`
- On the store picker, tap `+ Add a store`
- Tap `Create a new store` --> an in-progress modal should be shown briefly, then dismissed. the store name form is then presented
- Enter some store name and tap `Continue` when ready --> the store name should be used for the initial domain query in the next screen
- Search for another domain if you like, and select one domain to `Continue` --> a free Simple site should be created, and a store summary screen is shown with the store name and site slug
- Tap `Continue to payment` --> the eCommerce plan should be shown
- Tap `Create Store for $69.99/month` --> an in-progress screen should be shown about creating the store. after the site becomes available in the `/me/sites` response, the success screen should be shown
- Tap `Manage My Store` to continue to the new site in the app --> a WC version alert is probably shown because it's a Simple site without WC

### Screenshots



https://user-images.githubusercontent.com/1945542/202679805-2d0048c6-1ceb-4c18-b3ef-ea08f61d7911.mov



---
